File: src/MusicalConductor.Avalonia/MusicalConductor.Core/Monitoring/StatisticsManager.cs
using Microsoft.Extensions.Logging;

namespace MusicalConductor.Core.Monitoring;

/// <summary>
/// StatisticsManager - Performance metrics and statistics tracking
/// Handles all conductor statistics, performance metrics, and queue analytics
/// </summary>
public class StatisticsManager
{
    public class ConductorStatistics
    {
        public int TotalSequencesExecuted { get; set; }
        public int TotalBeatsExecuted { get; set; }
        public double AverageExecutionTime { get; set; }
        public int TotalSequencesQueued { get; set; }
        public int CurrentQueueLength { get; set; }
        public int MaxQueueLength { get; set; }
        public double AverageQueueWaitTime { get; set; }
        public int ErrorCount { get; set; }
        public double SuccessRate { get; set; }
        public double? LastExecutionTime { get; set; }
        public double SequenceCompletionRate { get; set; }
        public int ChainedSequences { get; set; }
    }

    private readonly ILogger<StatisticsManager> _logger;
    private ConductorStatistics _statistics;

    public StatisticsManager(ILogger<StatisticsManager> logger)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _statistics = new ConductorStatistics
        {
            TotalSequencesExecuted = 0,
            TotalBeatsExecuted = 0,
            AverageExecutionTime = 0,
            TotalSequencesQueued = 0,
            CurrentQueueLength = 0,
            MaxQueueLength = 0,
            AverageQueueWaitTime = 0,
            ErrorCount = 0,
            SuccessRate = 0,
            LastExecutionTime = null,
            SequenceCompletionRate = 0,
            ChainedSequences = 0
        };
    }

    /// <summary>
    /// Record a sequence execution
    /// </summary>
    public void RecordSequenceExecution(double executionTime)
    {
        _statistics.TotalSequencesExecuted++;

        // Update average execution time using exponential moving average
        const double alpha = 0.1; // Smoothing factor
        _statistics.AverageExecutionTime =
            _statistics.AverageExecutionTime * (1 - alpha) +
            executionTime * alpha;

        // Update success rate
        UpdateSuccessRate();

        _logger.LogDebug(
            "📊 StatisticsManager: Recorded sequence execution ({ExecutionTime:F2}ms)",
            executionTime);
    }

    /// <summary>
    /// Record a beat execution
    /// </summary>
    public void RecordBeatExecution()
    {
        _statistics.TotalBeatsExecuted++;
    }

    /// <summary>
    /// Record an error occurrence
    /// </summary>
    public void RecordError()
    {
        _statistics.ErrorCount++;
        UpdateSuccessRate();

        _logger.LogWarning("📊 StatisticsManager: Recorded error occurrence");
    }

    /// <summary>
    /// Record a sequence being queued
    /// </summary>
    public void RecordSequenceQueued()
    {
        _statistics.TotalSequencesQueued++;
        _statistics.CurrentQueueLength++;
        _statistics.MaxQueueLength = Math.Max(
            _statistics.MaxQueueLength,
            _statistics.CurrentQueueLength);
    }

    /// <summary>
    /// Record a sequence being dequeued
    /// </summary>
    public void RecordSequenceDequeued()
    {
        if (_statistics.CurrentQueueLength > 0)
        {
            _statistics.CurrentQueueLength--;
        }
    }

    /// <summary>
    /// Update queue wait time statistics
    /// </summary>
    public void UpdateQueueWaitTime(double waitTime)
    {
        // Simple moving average calculation
        const double alpha = 0.1; // Smoothing factor
        _statistics.AverageQueueWaitTime =
            _statistics.AverageQueueWaitTime * (1 - alpha) + waitTime * alpha;
    }

    /// <summary>
    /// Update success rate calculation
    /// </summary>
    private void UpdateSuccessRate()
    {
        var totalAttempts = _statistics.TotalSequencesExecuted + _statistics.ErrorCount;
        if (totalAttempts > 0)
        {
            _statistics.SuccessRate =
                ((double)_statistics.TotalSequencesExecuted / totalAttempts) * 100;
        }
        else
        {
            _statistics.SuccessRate = 100; // No attempts yet, assume 100%
        }
    }

    /// <summary>
    /// Get current statistics
    /// </summary>
    public ConductorStatistics GetStatistics()
    {
        return new ConductorStatistics
        {
            TotalSequencesExecuted = _statistics.TotalSequencesExecuted,
            TotalBeatsExecuted = _statistics.TotalBeatsExecuted,
            AverageExecutionTime = _statistics.AverageExecutionTime,
            TotalSequencesQueued = _statistics.TotalSequencesQueued,
            CurrentQueueLength = _statistics.CurrentQueueLength,
            MaxQueueLength = _statistics.MaxQueueLength,
            AverageQueueWaitTime = _statistics.AverageQueueWaitTime,
            ErrorCount = _statistics.ErrorCount,
            SuccessRate = _statistics.SuccessRate,
            LastExecutionTime = _statistics.LastExecutionTime,
            SequenceCompletionRate = _statistics.SequenceCompletionRate,
            ChainedSequences = _statistics.ChainedSequences
        };
    }

    /// <summary>
    /// Get enhanced statistics with additional metrics
    /// </summary>
    public object GetEnhancedStatistics(int mountedPlugins)
    {
        return new
        {
            Statistics = GetStatistics(),
            MountedPlugins = mountedPlugins
        };
    }

    /// <summary>
    /// Reset all statistics
    /// </summary>
    public void Reset()
    {
        _statistics = new ConductorStatistics
        {
            TotalSequencesExecuted = 0,
            TotalBeatsExecuted = 0,
            AverageExecutionTime = 0,
            TotalSequencesQueued = 0,
            CurrentQueueLength = 0,
            MaxQueueLength = 0,
            AverageQueueWaitTime = 0,
            ErrorCount = 0,
            SuccessRate = 0,
            LastExecutionTime = null,
            SequenceCompletionRate = 0,
            ChainedSequences = 0
        };

        _logger.LogInformation("🧹 StatisticsManager: All statistics reset");
    }

    /// <summary>
    /// Get performance summary
    /// </summary>
    public object GetPerformanceSummary()
    {
        var totalAttempts = _statistics.TotalSequencesExecuted + _statistics.ErrorCount;
        var errorRate = totalAttempts > 0
            ? ((double)_statistics.ErrorCount / totalAttempts) * 100
            : 0;

        return new
        {
            ExecutionEfficiency = _statistics.SuccessRate,
            QueueEfficiency = _statistics.AverageQueueWaitTime > 0
                ? Math.Max(0, 100 - _statistics.AverageQueueWaitTime / 1000)
                : 100,
            ErrorRate = errorRate,
            Throughput = _statistics.AverageExecutionTime > 0
                ? 1000 / _statistics.AverageExecutionTime  // sequences per second
                : 0
        };
    }

    /// <summary>
    /// Get queue analytics
    /// </summary>
    public object GetQueueAnalytics()
    {
        return new
        {
            CurrentLoad = _statistics.CurrentQueueLength,
            MaxLoadReached = _statistics.MaxQueueLength,
            AverageWaitTime = _statistics.AverageQueueWaitTime,
            TotalProcessed = _statistics.TotalSequencesExecuted
        };
    }

    /// <summary>
    /// Log current statistics
    /// </summary>
    public void LogStatistics()
    {
        _logger.LogInformation(
            "📊 Conductor Statistics: Sequences={TotalSequences} | Beats={TotalBeats} | AvgTime={AvgTime:F2}ms | Success={SuccessRate:F1}% | Errors={Errors}",
            _statistics.TotalSequencesExecuted,
            _statistics.TotalBeatsExecuted,
            _statistics.AverageExecutionTime,
            _statistics.SuccessRate,
            _statistics.ErrorCount);
    }

    /// <summary>
    /// Get debug information
    /// </summary>
    public object GetDebugInfo()
    {
        return new
        {
            Statistics = _statistics,
            PerformanceSummary = GetPerformanceSummary(),
            QueueAnalytics = GetQueueAnalytics()
        };
    }
}

File: packages/musical-conductor/modules/communication/sequences/monitoring/StatisticsManager.ts
/**
 * StatisticsManager - Performance metrics and statistics tracking
 * Handles all conductor statistics, performance metrics, and queue analytics
 */

import type { ConductorStatistics } from "../SequenceTypes.js";

export class StatisticsManager {
  private statistics: ConductorStatistics;

  constructor() {
    this.statistics = {
      totalSequencesExecuted: 0,
      totalBeatsExecuted: 0,
      averageExecutionTime: 0,
      totalSequencesQueued: 0,
      currentQueueLength: 0,
      maxQueueLength: 0,
      averageQueueWaitTime: 0,
      errorCount: 0,
      successRate: 0,
      lastExecutionTime: null,
      sequenceCompletionRate: 0,
      chainedSequences: 0,
    };
  }

  /**
   * Record a sequence execution
   * @param executionTime - Execution time in milliseconds
   */
  recordSequenceExecution(executionTime: number): void {
    this.statistics.totalSequencesExecuted++;

    // Update average execution time using exponential moving average
    const alpha = 0.1; // Smoothing factor
    this.statistics.averageExecutionTime =
      this.statistics.averageExecutionTime * (1 - alpha) +
      executionTime * alpha;

    // Update success rate
    this.updateSuccessRate();

    (globalThis as any).__MC_LOG(
      `📊 StatisticsManager: Recorded sequence execution (${executionTime.toFixed(
        2
      )}ms)`
    );
  }

  /**
   * Record a beat execution
   */
  recordBeatExecution(): void {
    this.statistics.totalBeatsExecuted++;
  }

  /**
   * Record an error occurrence
   */
  recordError(): void {
    this.statistics.errorCount++;
    this.updateSuccessRate();

    (globalThis as any).__MC_WARN("📊 StatisticsManager: Recorded error occurrence");
  }

  /**
   * Record a sequence being queued
   */
  recordSequenceQueued(): void {
    this.statistics.totalSequencesQueued++;
    this.statistics.currentQueueLength++;
    this.statistics.maxQueueLength = Math.max(
      this.statistics.maxQueueLength,
      this.statistics.currentQueueLength
    );
  }

  /**
   * Record a sequence being dequeued
   */
  recordSequenceDequeued(): void {
    if (this.statistics.currentQueueLength > 0) {
      this.statistics.currentQueueLength--;
    }
  }

  /**
   * Update queue wait time statistics
   * @param waitTime - Wait time in milliseconds
   */
  updateQueueWaitTime(waitTime: number): void {
    // Simple moving average calculation
    const alpha = 0.1; // Smoothing factor
    this.statistics.averageQueueWaitTime =
      this.statistics.averageQueueWaitTime * (1 - alpha) + waitTime * alpha;
  }

  /**
   * Update success rate calculation
   */
  private updateSuccessRate(): void {
    const totalAttempts =
      this.statistics.totalSequencesExecuted + this.statistics.errorCount;
    if (totalAttempts > 0) {
      this.statistics.successRate =
        (this.statistics.totalSequencesExecuted / totalAttempts) * 100;
    } else {
      this.statistics.successRate = 100; // No attempts yet, assume 100%
    }
  }

  /**
   * Get current statistics
   * @returns Current conductor statistics
   */
  getStatistics(): ConductorStatistics {
    return { ...this.statistics };
  }

  /**
   * Get enhanced statistics with additional metrics
   * @param mountedPlugins - Number of mounted plugins
   * @returns Enhanced statistics
   */
  getEnhancedStatistics(
    mountedPlugins: number
  ): ConductorStatistics & { mountedPlugins: number } {
    return {
      ...this.statistics,
      mountedPlugins,
    };
  }

  /**
   * Reset all statistics
   */
  reset(): void {
    this.statistics = {
      totalSequencesExecuted: 0,
      totalBeatsExecuted: 0,
      averageExecutionTime: 0,
      totalSequencesQueued: 0,
      currentQueueLength: 0,
      maxQueueLength: 0,
      averageQueueWaitTime: 0,
      errorCount: 0,
      successRate: 0,
      lastExecutionTime: null,
      sequenceCompletionRate: 0,
      chainedSequences: 0,
    };

    (globalThis as any).__MC_LOG("🧹 StatisticsManager: All statistics reset");
  }

  /**
   * Get performance summary
   * @returns Performance summary object
   */
  getPerformanceSummary(): {
    executionEfficiency: number;
    queueEfficiency: number;
    errorRate: number;
    throughput: number;
  } {
    const totalAttempts =
      this.statistics.totalSequencesExecuted + this.statistics.errorCount;
    const errorRate =
      totalAttempts > 0
        ? (this.statistics.errorCount / totalAttempts) * 100
        : 0;

    return {
      executionEfficiency: this.statistics.successRate,
      queueEfficiency:
        this.statistics.averageQueueWaitTime > 0
          ? Math.max(0, 100 - this.statistics.averageQueueWaitTime / 1000)
          : 100,
      errorRate,
      throughput:
        this.statistics.averageExecutionTime > 0
          ? 1000 / this.statistics.averageExecutionTime
          : 0, // sequences per second
    };
  }

  /**
   * Get queue analytics
   * @returns Queue performance analytics
   */
  getQueueAnalytics(): {
    currentLoad: number;
    maxLoadReached: number;
    averageWaitTime: number;
    totalProcessed: number;
  } {
    return {
      currentLoad: this.statistics.currentQueueLength,
      maxLoadReached: this.statistics.maxQueueLength,
      averageWaitTime: this.statistics.averageQueueWaitTime,
      totalProcessed: this.statistics.totalSequencesQueued,
    };
  }

  /**
   * Check if performance thresholds are exceeded
   * @returns Performance warnings
   */
  getPerformanceWarnings(): string[] {
    const warnings: string[] = [];

    if (this.statistics.averageExecutionTime > 5000) {
      warnings.push("High average execution time (>5s)");
    }

    if (this.statistics.averageQueueWaitTime > 10000) {
      warnings.push("High average queue wait time (>10s)");
    }

    if (this.statistics.successRate < 95) {
      warnings.push("Low success rate (<95%)");
    }

    if (this.statistics.currentQueueLength > 10) {
      warnings.push("High current queue length (>10)");
    }

    if (this.statistics.errorCount > 100) {
      warnings.push("High error count (>100)");
    }

    return warnings;
  }

  /**
   * Export statistics for external monitoring
   * @returns Statistics in monitoring format
   */
  exportForMonitoring(): Record<string, number> {
    return {
      "conductor.sequences.executed": this.statistics.totalSequencesExecuted,
      "conductor.beats.executed": this.statistics.totalBeatsExecuted,
      "conductor.execution.avg_time_ms": this.statistics.averageExecutionTime,
      "conductor.queue.length": this.statistics.currentQueueLength,
      "conductor.queue.max_length": this.statistics.maxQueueLength,
      "conductor.queue.avg_wait_time_ms": this.statistics.averageQueueWaitTime,
      "conductor.errors.count": this.statistics.errorCount,
      "conductor.success.rate_percent": this.statistics.successRate,
    };
  }

  /**
   * Get debug information
   * @returns Debug statistics information
   */
  getDebugInfo(): {
    statistics: ConductorStatistics;
    performanceSummary: ReturnType<StatisticsManager["getPerformanceSummary"]>;
    queueAnalytics: ReturnType<StatisticsManager["getQueueAnalytics"]>;
    warnings: string[];
  } {
    return {
      statistics: this.getStatistics(),
      performanceSummary: this.getPerformanceSummary(),
      queueAnalytics: this.getQueueAnalytics(),
      warnings: this.getPerformanceWarnings(),
    };
  }
}
